### PR TITLE
gh-119823: support bytes/bytearray in complex() constructor

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-119823

### CODE REVIEW

File: Lib/test/test_complex.py [hunk 1]
```diff
@@ -632,6 +632,22 @@ def test_constructor_negative_nans_from_string(self):
...
+    def test_other_non_numeric_input_types(self):
+        class CustomStr(str): ...
+        class CustomBytes(bytes): ...
+        class CustomByteArray(bytearray): ...
+
+        factories = [bytes, bytearray, lambda b: CustomStr(b.decode()),
+                     CustomBytes, CustomByteArray]
+
+        for f in factories:
+            x = f(b"(1+1j)")
```
Comment (sobolevn) [inline Lib/test/test_complex.py]: this is a part of `subTest`

File: Objects/complexobject.c [hunk 2]
```diff
@@ -941,6 +941,17 @@ actual_complex_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
     if (PyUnicode_Check(arg)) {
         return complex_subtype_from_string(type, arg);
     }
+    else if (PyByteArray_Check(arg) || PyBytes_Check(arg)) {
+        if (PyByteArray_Check(arg)) {
+            const char *string = PyByteArray_AS_STRING(arg);
```
Comment (sobolevn) [inline Objects/complexobject.c]: Pls, note that there are multiple warnings

Comment (skirpichev) [reply Objects/complexobject.c]: Yeah, I learned a polite way to say "this code so broken, that will compile just on gcc".

File: Lib/test/test_complex.py (hunk 1 above)
Comment (skirpichev) [reply Lib/test/test_complex.py]: Now factory value is used as a parameter.

**Review by sobolevn — COMMENTED** [review]
Two questions:
1. Shouldn't you use `PyBytes_GET_SIZE` and `PyByteArray_GET_SIZE` here instead of `Py_SIZE`?
2. Should this be documented?

**Comment by sobolevn** [comment]
Anwsering 1.

they are kinda the same:

```c
static inline Py_ssize_t PyByteArray_GET_SIZE(PyObject *op) {
    PyByteArrayObject *self = _PyByteArray_CAST(op);
    return Py_SIZE(self);
}
```

**Comment by skirpichev** [comment]
1) yes, probably you are right.
2) probably, no - it might be (highly unlikely, but) an implementation detail.  BTW, I've seen just one place (int constructor), where Py_SIZE is used this way.

**Pull request closed without merging**